### PR TITLE
Clean-up Geant4 sensitive detectors3

### PR DESCRIPTION
**Pull request by civanch**
will test after pre3

### CODE REVIEW

File: SimG4CMS/Calo/interface/CaloSD.h [hunk 1]
```diff
@@ -21,14 +21,12 @@
...
 #include "FWCore/MessageLogger/interface/MessageLogger.h"
-
-// To be replaced by something else 
-/* #include "Utilities/Notification/interface/TimerProxy.h" */
  
 #include "G4VPhysicalVolume.hh"
 #include "G4Track.hh"
 #include "G4VGFlashSensitiveDetector.hh"
 
+#include <boost/cstdint.hpp>
```
Comment (kpedro88) [inline SimG4CMS/Calo/interface/CaloSD.h]: can this just be `<cstdint>`?

File: SimG4CMS/Forward/src/ZdcSD.cc [hunk 2]
```diff
@@ -260,8 +247,8 @@ double ZdcSD::getEnergyDeposit(const G4Step * aStep, edm::ParameterSet const & p
       float thFullRefl = 23.;
       float thFullReflRad = thFullRefl*pi/180.;
 
-      edm::ParameterSet m_ZdcSD = p.getParameter<edm::ParameterSet>("ZdcSD");
-      thFibDir  = m_ZdcSD.getParameter<double>("FiberDirection");
+      //edm::ParameterSet m_ZdcSD = p.getParameter<edm::ParameterSet>("ZdcSD");
```
Comment (kpedro88) [inline SimG4CMS/Forward/src/ZdcSD.cc]: remove commented-out code

File: SimG4CMS/Calo/src/ECalSD.cc [hunk 3]
```diff
@@ -286,63 +262,41 @@ uint16_t ECalSD::getDepth(const G4Step * aStep) {
...
-      std::string lvname = lv->GetName();
-      int k1 = (lvname.find("EFRY")!=std::string::npos) ? 2 : 0;
-      int k2 = (lvname.find("refl")!=std::string::npos) ? 1 : 0;
-      int kk = k1+k2;
-      double rz = (k1 == 0) ? (hitPoint->GetPosition()).rho() : 
-        std::abs((hitPoint->GetPosition()).z());
-      edm::LogVerbatim("EcalSim") << lvname << " # " << k1 << ":" << k2 << ":" 
-                                  << kk << " rz " << rz << " D " << thisX0;
-      g2L_[kk]->Fill(rz,thisX0);
+    std::string lvname = lv->GetName();
```
Comment (kpedro88) [inline SimG4CMS/Calo/src/ECalSD.cc]: could this be a const reference?

File: SimG4CMS/Calo/src/HCalSD.cc [hunk 4]
```diff
@@ -564,7 +570,7 @@ uint32_t HCalSD::setDetUnitId(const G4Step * aStep) {
 void HCalSD::setNumberingScheme(HcalNumberingScheme * scheme) {
   if (scheme != nullptr) {
     edm::LogInfo("HcalSim") << "HCalSD: updates numbering scheme for " << GetName();
-    if (numberingScheme) delete numberingScheme;
+    delete numberingScheme;
```
Comment (VinInn) [inline SimG4CMS/Calo/src/HCalSD.cc]: use unique_ptr<HcalNumberingScheme>

File: SimG4CMS/HGCalTestBeam/src/HGCalTB16SD01.cc [hunk 5]
```diff
@@ -44,9 +44,9 @@ HGCalTB16SD01::HGCalTB16SD01(const std::string& name, const DDCompactView & cpv,
 
 HGCalTB16SD01::~HGCalTB16SD01() {}
```
Comment (VinInn) [inline SimG4CMS/HGCalTestBeam/src/HGCalTB16SD01.cc]: ~HGCalTB16SD01() = default
in the header

File: SimG4CMS/HcalTestBeam/src/HcalTB02SD.cc [hunk 6]
```diff
@@ -84,12 +84,12 @@ HcalTB02SD::~HcalTB02SD() {
...
  
-double HcalTB02SD::getEnergyDeposit(G4Step * aStep) {
+double HcalTB02SD::getEnergyDeposit(const G4Step * aStep, bool&) {
   
   if (aStep == nullptr) {
     return 0;
   } else {
-    preStepPoint        = aStep->GetPreStepPoint();
+    const G4StepPoint* preStepPoint        = aStep->GetPreStepPoint();
     G4String nameVolume = preStepPoint->GetPhysicalVolume()->GetName();
```
Comment (VinInn) [inline SimG4CMS/HcalTestBeam/src/HcalTB02SD.cc]: auto const &

File: SimG4CMS/HGCalTestBeam/plugins/AHCalSD.cc [hunk 7]
```diff
@@ -40,7 +40,7 @@ AHCalSD::AHCalSD(const std::string& name, const DDCompactView & cpv,
 
 AHCalSD::~AHCalSD() { }
```
Comment (VinInn) [inline SimG4CMS/HGCalTestBeam/plugins/AHCalSD.cc]: default in the header

File: SimG4CMS/Forward/src/ZdcSD.cc [hunk 8]
```diff
@@ -416,14 +403,13 @@ void ZdcSD::setNumberingScheme(ZdcNumberingScheme* scheme) {
   if (scheme != nullptr) {
     edm::LogInfo("ForwardSim") << "ZdcSD: updates numbering scheme for " 
 			       << GetName();
-    if (numberingScheme) delete numberingScheme;
+    delete numberingScheme;
     numberingScheme = scheme;
```
Comment (VinInn) [inline SimG4CMS/Forward/src/ZdcSD.cc]: use unique_ptr
maybe all those commonality among call SD should be consolidated into a (template?) base class

File: SimG4CMS/Forward/src/ZdcSD.cc [hunk 9]
```diff
@@ -55,16 +54,13 @@ ZdcSD::ZdcSD(const std::string& name, const DDCompactView & cpv,
   
   if(useShowerLibrary){
     showerLibrary = new ZdcShowerLibrary(name, cpv, p);
```
Comment (VinInn) [inline SimG4CMS/Forward/src/ZdcSD.cc]: use unique_ptr
make destructor default

File: SimG4CMS/Forward/src/CastorSD.cc [hunk 10]
```diff
@@ -100,9 +102,9 @@ double CastorSD::getEnergyDeposit(G4Step * aStep) {
...
   
-  G4StepPoint*       preStepPoint = aStep->GetPreStepPoint();
-  G4VPhysicalVolume* currentPV    = preStepPoint->GetPhysicalVolume();
-  G4LogicalVolume*   currentLV    = currentPV->GetLogicalVolume();
+  const G4StepPoint*       preStepPoint = aStep->GetPreStepPoint();
+  const G4VPhysicalVolume* currentPV    = preStepPoint->GetPhysicalVolume();
+  const G4LogicalVolume*   currentLV    = currentPV->GetLogicalVolume();
 
 #ifdef debugLog
   G4String           name   = currentPV->GetName();
```
Comment (VinInn) [inline SimG4CMS/Forward/src/CastorSD.cc]: all those G4String on the left should become ```auto const &```

**Comment by VinInn** [comment]
I see a lot of duplicate code.
It shall be factorized, so maintenance is easier

**Comment by civanch** [comment]
@kpedro88, @VinInn , thanks for comments, I will try to implement, this process of code cleanup was started partially because of very big code duplication and also some inefficiency in the code is present. It is a bit non-trivial, because of historical mess, so I am trying to do via iterations.

**Comment by civanch** [comment]
This PR is closed and #21797 is created with comments implemented.

**Pull request closed without merging**